Allow AIX to use openxlC as the C compiler

FractalBoy opened the pull request:
The recommended compiler for AIX is now the openxlC compiler (ibm-clang).

This PR updates AIX hints so that it can be used. Additionally, it fixes an issue with Perl's implementation of Clang TSA so that the perl_tsa_mutex_* functions are accessible to ext/re. This was required in order for AIX to build cleanly.

We also remove NEED_PTHREAD_INIT, since pthread_init doesn't exist on AIX (nor anywhere from what I can tell). https://www.ibm.com/docs/en/aix/7.3.0?topic=options-supported-interfaces

I'm also updating the AIX documentation to recommend ibm-clang_r over cc/xlc.

Addresses #23333


---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

FractalBoy committed:
Update AIX README

FractalBoy committed:
Update perldelta

FractalBoy committed:
Update AIX hints to default to ibm-clang_r if found

Leont commented:
I would recommend running the tests before sending an PR ;-).

You need to run the `Porting/updateAUTHORS.pl` to add yourself to the `AUTHORS` file.

You also need to run `make regen` and commit the results of that after editing `embed.fnc`.

And you need to reformat some verbatim sections `perlaix.pod` to not be too long.

FractalBoy committed:
Update AIX hints to allow clang to work

FractalBoy committed:
Add perl_tsa_mutex_* to exported functions

When PERL_TSA_ACTIVE is defined, perl_tsa_mutex_lock and perl_tsa_mutex_unlock
are required to be defined. However, they are not exported so they cannot be used
by core extensions such as re.

FractalBoy committed:
Remove NEED_PTHREAD_INIT macro and its usage in AIX hints

pthread_init is not a real function (in POSIX or AIX pthread), and
doesn't appear to have ever existed. Trying to use it causes a compiler
failure when using ibm-clang.

FractalBoy commented:
Okay, should be good now.

* I'm already in the AUTHORS file but I had some commits where the author/committer used a different email address
* All porting.t tests are fixed
* The original change to embed.fnc didn't work because perl_tsa_mutex_{un,}lock are already defined in perl.h. Adding it to embed.fnc without defining it in perl.h broke Windows builds. Instead, I updated perl.h so that they are always defined if USE_ITHREADS and I_PTHREAD are defined, and updated makedef.pl to ensure they are exported. This is consistent with other stuff defined in perl.h.

bulk88 commented:
> * The original change to embed.fnc didn't work because perl_tsa_mutex_{un,}lock are already defined in perl.h. Adding it to embed.fnc without defining it in perl.h broke Windows builds. Instead, I updated perl.h so that they are always defined if USE_ITHREADS and I_PTHREAD are defined, and updated makedef.pl to ensure they are exported. This is consistent with other stuff defined in perl.h.

perl.h isn't supposed to know the actual "types" of function symbols/function pointers. types, decls, attrs that cross TUs.

```
LEFT_SIDE void
Perl_xyz(pTHX- SV* sv)
RIGHT_SIDE ;
```
Only embed.fnc/embed.h/proto.h/makedef.pl know what LEFT_SIDE and RIGHT_SIDE expand to.

Only embed.fnc/embed.h/proto.h know how to correctly apply the no_return optimization on all CPUs/OSes safely.

https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms#Overview


>A pointer to a function declared as extern “C” isn’t interchangeable with a function declared as extern “c++”. This behavior differs from the ARM64 ABI, in which the functions are interchangeable.

>For more information about the generic ARM64 C++ ABI, see “C++ Application Binary Interface Standard for the ARM 64-bit architecture” at developer.arm.com.

What is a calling convention? (Visual C has 4@i386, 2@ArmCE32, 2~4@ArmNT32, 2 for Win64 any CPU. Intel C for Win32-i386 has 5-6. Clang for Win32-i386 >= 8. Linux has SysV aka _cdecl and __syscall.

ARM LLC has licensed/allocated in their symbolic debugger file format, ranges of numbers of the "calling convention" struct member, to dozens of independent ARM targetted C compilers. And those C compilers in theory now can create "random" calling conventions that /bin/gdb must decode from the symbol files.  This list is mostly for using -O0 non-inlined "__builtin_gcc_arm32_memcmp()" and "__builtin_intel_arm32_memcmp()" function. The list is:

-aeabi (in-house)
-anonXyz Guaranteed not to clash with any registered vendor,
-cxa C++pseudo-vendor(in-house),
-FSL Freescale Semiconductor
-GHS Green Hills Systems
-gnu Free Software Foundation
-iar IAR Systems
-intel Intel Corporation
-ixs Intel Xscale
-llvm The LLVM/Clang projects
-PSI PalmSource
-RAL Rowley Associates
-TASKING Altium
-TI TI Inc,
-tls Reserved for use in thread-local storage routines
-WRS Wind River Systems

This is hopeless for having ABI linktime compatibility between 2 .o files from 2 different CC vendors. Although nobody has a sane need to do this, you can mix-n-match ARM toolchains and mix-n-match ARM CCs and win the grame if you follow all the rules and spend an eternity settingtthe right bespoke cmg line largs on the CC cmd line, and a ton of complicated func attributes to the declaration in the .c file, and then finally the the alien 3rd party out-of-ecosystem linker  will accept your .op files. Best to avoid this always.

What is a function pointer? (a 128 bit memory address on MIPS, Z/OS, and Itanium, but for legacy apps that turn on a legacy emulation mode they will get 64-bit data ptr (to malloc blocks) to a 128 bit C integers stored in a malloc block that was specially made for them, and only then once the cpu read all 128 bits, will control flow transfer out of your func body to you agree.

What is `extern` without dbl quote "C" on a random unknown age ( OS *  CC * CPU arch * annual new CPU opcodes )?
How do 32 byte pass by copy C structs as arg 1 in C++ behave in relation to where the `this` or if there was no `this` pointer in this static class method?

Are you iOS ARM or OSX ARM?
```
EXTERN_C int perl_tsa_mutex_lock(perl_mutex* mutex)
  PERL_TSA_ACQUIRE(*mutex)
  PERL_TSA_NO_TSA;
EXTERN_C int perl_tsa_mutex_unlock(perl_mutex* mutex)
  PERL_TSA_RELEASE(*mutex)
  PERL_TSA_NO_TSA;
#elif defined(USE_ITHREADS) && defined(I_PTHREAD)
EXTERN_C int perl_tsa_mutex_lock(perl_mutex* mutex);
EXTERN_C int perl_tsa_mutex_unlock(perl_mutex* mutex);
#endif
```
Thisa above  is the wrong way to right it on Win32/Win64/all .def file using OSes. "EXTERN_C"` means the function will be  reachable to all TUs inside of `perl5xx.dll` and no where.  "EXTERN_C" tag will not put that function symbol into the Windows PE export symbol table.Just call `EXTERN_C` functions by their real names, "static declared C function that can be exposed to all of its consumers  in all TUs inside 1 .dll or 1.exe binar, but these Extern_C symbols can't escape the binary, unless someone captures theior fucntion ptr, and uses alternative I/O means to move that ptr to a vtable to some other consumer lib, 

declspec( __dllexport ) means the function is not "static declared inside 1 .exe/.dll" and is now globally declared and visible between 2 separate diskfiles with .dll ends. Similar thing for __dllimport, that means the Assembler have to write a variant of  the x86  call does 1 memory before transfering control since its cross dll boundaries. Same binary calls dont do the memory read, they know the hard coded offsets on how much forwards and backwards they need to jump from current RIP register. This is unlike unix where inside-the-same binary still have do a memory read and indirect through the PLT/GOT.  Windows only uses the its analog of the PLY GOT to cross disk files, Linux use it for intra-self-.so/dll calls unlike windows, and for window's inter different .so/.dll file jumps to.

So basically Win32 .C while its at `".i"` phase code always need to know if the current stubs/predeclareds are comming from another .dll at link phase, or coming from MY/our .dll at link phase,

example of a double or extra read Win32 __dllimport call,
```
DWORD count;
void *fn = *(void**)(RIP+0x45204));
BOOL = (( BOOL (__stdcall *)(LPCSTR, LPCSTR, DWORD)) fn)("hello", "world", &count);
```

embed.fnc and makedef.dll know how to tag these declarations for internal libperl use, or tag them for outside world use (CPAN XS), DIYing their logic in perl.h always causes a disaster because your creating function pointers/symbols with the wrong calling convetions.

This is how good code from embed.fnc.
```
PERL_CALLCONV SV *
Perl_newSV_true(pTHX)
        __attribute__warn_unused_result__;
#define PERL_ARGS_ASSERT_NEWSV_TRUE

PERL_CALLCONV_NO_RET void
Perl_noperl_die(const char *pat, ...)
        __attribute__noreturn__
        __attribute__format__(__printf__,1,2);
#define PERL_ARGS_ASSERT_NOPERL_DIE             \
        assert(pat)
```

FractalBoy commented:
I'm open to suggestions on how to do it correctly. I just hacked it until it worked on AIX and didn't break the builds - I fully expected to get some feedback before it got merged.

khwilliamson commented:
It did not appear feedback was going to be forthcoming, so I merged it.  It's early enough in the development cycle that problems that may arise can be handled.

But I do not understand what @bulk88 is saying.  He needs to rewrite it so that there are clearly understandable actionable steps we can take to address whatever concerns he has.
